### ISSUE TITLE
Fail with error (rather than hanging) when init_op requires starting queues (ManagedSesssion etc)

### ISSUE THREAD

**Issue by elibixby** [tensorflow/tensorflow#6624]
In 0.12 and in master

`MonitoredSession` wraps the original `SessionCreator` in a `_CoordinatedSessionCreator` in an attempt to ensure queue runners are started before the session is run, however the original session has to be created first (as it is an argument to start_queue_runner). In the case of `ChiefSessionCreator`, `ChiefSessionCreator.create_session` calls `ChiefSessionCreator._get_session_manager`
which instantiates a new `SessionManager` and returns it. Then `ChiefSessionCreator.create_session` calls `SessionManager.prepare_session` which calls `sess.run(init_op)`. Thus resulting in `sess.run(init_op)` being run before `start_queue_runner` is called. 

This makes it impossible to initialize variables from queues. Even if this is intended behavior, it causes the initialization of `MonitoredSession` to stall without any logging output, and not respond to SIGTERM.

It seems that `prepare_session` may need to be split into two calls for this use case.

**Comment by mrry** [comment]
Unless I'm mistaken, it's quite important for initialization to happen before starting the queue runners, because [`input_producer()`](https://github.com/tensorflow/tensorflow/blob/7c36309c37b04843030664cdc64aca2bb7d6ecaa/tensorflow/python/training/input.py#L101) calls [`limit_epochs()`](https://github.com/tensorflow/tensorflow/blob/7c36309c37b04843030664cdc64aca2bb7d6ecaa/tensorflow/python/training/input.py#L69), which sometimes creates a variable... that needs to be initialized before starting the queue runner (for the input producer).

What's the use case for initializing a variable from a queue? It sounds quite exotic, so perhaps it would be possible to achieve with a workaround in the user program.

**Comment by elibixby** [comment]
Ahh shoot. You're right. I was just testing my code with constants upstream from the enqueue_ops (and initializing from a queue worked fine, because I could start the queue before running the init_op) but if I put variables upstream from the enqueue ops it throws a failed precondition. 

Going to rescope this bug to failing more noisily when unstarted queues are inputs to ops. FYI my exotic use case was just a regular old rolling window pattern: https://gist.github.com/elibixby/cedd3be67020169ff3efc47dbb603af6 

I'll try to find another way to do it.

**Comment by LMescheder** [comment]
I had the same issue when I wanted to use data dependent initialization in my neural net. I general, data dependent initialization is not an exotic thing to do, is it? Is there a better workaround than using assign ops after initialization is completed?

**Comment by girving** [comment]
@elibixby What's the status of this issue?

**Comment by elibixby** [comment]
Not sure there is any fix in the works. But I believe there may be a workaround, where you can create a separate GraphKey for variables that are downstream of Queues, and run initialization of those variables after queue initialization. Have to test it but believe it should work.

**Comment by girving** [comment]
Actually, did https://github.com/tensorflow/tensorflow/commit/b25d1c7d3e9f30925aa132ba62e79d281191a3dc fix this?  It makes `Variable.initialized_value` not reinitialize already initialized variables, which should make data dependent initialization quite a bit easier.

**Comment by elibixby** [comment]
It's possible, busy today but I can test tomorrow if no-one beats me to it.

**Comment by elibixby** [comment]
@girving Yep! Good catch! This use case is now possible, if you remember to start queue runners, *however*, it still hangs indefinitely if you do not start queue runners. It should be possible for evaluation of graph nodes to determine if there are upstream queue nodes, and queues have not been started, which would help immensely in debugging this common error (whether it's worth the effort given `tf.contrib.data` is coming, is another question).

To be clear with the following graph:

```
import tensorflow as tf

graph = tf.Graph()
with graph.as_default():
   queue = tf.train.string_input_producer(['foo', 'bar', 'baz'])
   var = tf.Variable(queue.dequeue_many(3))
```

The following code works:

```
with tf.Session(graph=graph) as sess:
   tf.train.start_queue_runners(sess=sess)
   tf.global_variables_initializer().run()
   print(sess.run(var))
```

But the easy mistake of reversing two lines:

```
with tf.Session(graph=graph) as sess:
   tf.global_variables_initializer().run()
   tf.train.start_queue_runners(sess=sess)
   print(sess.run(var))
```

Results in a program that hangs indefinitely

**Comment by girving** [comment]
@mrry Do you think it's feasible to programmatically warn about `QueueRunners` not being started?  The warning could even point people to `tf.contrib.data`. :)

**Comment by mrry** [comment]
@girving I once hacked up something that added to each `tf.Tensor` the set of queue runners on which it depended, and printed a warning when you tried to evaluate a tensor that depended on a queue runner that hadn't been started. I think it foundered because people ignore warnings and want errors, but there was a concern about false positives (e.g. perhaps you could feed a value to avoid depending on the queue, or it could be control-flow sensitive, and it wasn't clear how much metadata we should track to avoid the problem).

**Comment by elibixby** [comment]
At the very least, I find it concerning that running of the initialization op doesn't respond to `SIGTERM` and only responds to `SIGKILL` when it's waiting for a queue dependency. What gives there?

**Comment by mrry** [comment]
[This Stack Overflow answer](https://stackoverflow.com/q/14707049/3574081) goes into some of the details. Short answer: invoking native code on the main thread of a Python process can prevent the signal from being delivered.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by michaelisard** [comment]
Given that I think queues are now mostly deprecated, should we mark this wontfix? If no response after a while, we will close this.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by alextp** [comment]
I think this is no longer relevant in the world of tf.data, so closing. Please reopen if it's still an issue.
